### PR TITLE
Cleanup of join code

### LINKED ISSUE #1858: Reenable fast matching for R != 3.3.0
https://github.com/hadley/dplyr/commit/716875978db16d6f621945dd758d0ccb9fa82452#diff-796101d8dc53d733ea4bf9ff4190e9fbR11

Follow-up to #1810.

### PR DESCRIPTION
**Pull request by krlmlr**
in preparation for #2033. Considerably simplifies string-string and string-factor joins.

Fixes #1858.

### CODE REVIEW

**Comment by krlmlr** [comment]
@hadley: This gives a ~3-4x speedup for joining strings, if they don't need to be reencoded.

download

**Comment by hadley** [comment]
Nice!

File: R/zzz.r [hunk 1]
```diff
@@ -1,11 +1,22 @@
...
     warning(
       "Installed Rcpp (", utils::packageVersion("Rcpp"), ") different from ",
       "Rcpp used to build dplyr (", Rcpp_version, ").\n",
-      "Please reinstall dplyr to avoid random crashes.",
+      "Please reinstall dplyr to avoid random crashes or undefined behavior.",
+      call. = FALSE
+    )
+  }
+
+  if (R_version != R.version.string) {
```
Comment (hadley) [inline R/zzz.r]: Do you remember why you added this check here?

Comment (krlmlr) [inline R/zzz.r]: We're checking the R version at compile time, currently only for a workaround for a problem in R 3.3.0 .

**Pull request merged**